### ISSUE TITLE
IndexError: tuple index out of range

### ISSUE THREAD

**Issue by Ant0ny44** [stnoah1/infogcn2#4]
Hi, thanks for your awesome work! Here occurs a problem when I try to train InfoGCN++ on NTU120 dataset. 
Here are some details:
```
Traceback (most recent call last):
  File "main.py", line 464, in <module>
    main()
  File "main.py", line 461, in main
    processor.start()
  File "main.py", line 428, in start
    self.train(epoch, save_model=save_model)
  File "main.py", line 241, in train
    y_hat, x_hat, z_0, z_hat, kl_div = self.model(x)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/infoGCN--/model/sode.py", line 244, in forward
    z_0, z_hat, z_hat_shifted = self.extrapolate(z, self.arange_n_step.to(z.dtype))
  File "/home/infoGCN--/model/sode.py", line 182, in extrapolate
    zs = self.diffeq_solver(z_0, t) # z_i = 2, (b t), c, v
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/infoGCN--/model/sode.py", line 36, in forward
    pred_y = odeint(self.ode_func, first_point, time_steps_to_predict,
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torchdiffeq/_impl/odeint.py", line 77, in odeint
    solution = solver.integrate(t)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torchdiffeq/_impl/solvers.py", line 105, in integrate
    dy, f0 = self._step_func(self.func, t0, dt, t1, y0)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torchdiffeq/_impl/fixed_grid.py", line 10, in _step_func
    f0 = func(t0, y0, perturb=Perturb.NEXT if self.perturb else Perturb.NONE)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torchdiffeq/_impl/misc.py", line 189, in forward
    return self.base_func(t, y)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/infoGCN--/model/sode.py", line 59, in forward
    x = self.conv1(x)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 313, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/torch/nn/modules/conv.py", line 309, in _conv_forward
    return F.conv1d(input, weight, bias, self.stride,
  File "/usr/local/miniconda3/lib/python3.8/site-packages/apex/amp/wrap.py", line 21, in wrapper
    args[i] = utils.cached_cast(cast_fn, args[i], handle.cache)
  File "/usr/local/miniconda3/lib/python3.8/site-packages/apex/amp/utils.py", line 97, in cached_cast
    if cached_x.grad_fn.next_functions[1][0].variable is not x:
IndexError: tuple index out of range
```
 The hyper parameters are as followed 
```
{'debug': False, 'log_dir': '.', 'model_saved_name': '', 'noise_ratio': 0.5, 'project': 'SODE', 'n_desired': 40000, 'num_point': 25, 'num_person': 2, 'num_class': 120, 'dataset': 'ntu120', 'datacase': 'NTU120_CSet', 'use_vel': False, 'phase': 'train', 'save_score': True, 'seed': 1, 'log_interval': 100, 'save_interval': 1, 'save_epoch': 0, 'eval_interval': 5, 'print_log': True, 'show_topk': [1, 5], 'feeder': 'feeders.feeder_ntu.Feeder', 'num_worker': 8, 'random_rot': True, 'repeat': 1, 'window_size': 64, 'base_channel': 64, 'weights': None, 'ignore_weights': [], 'n_heads': 3, 'depth': 4, 'k': 8, 'z_prior_gain': 3, 'graph': 'graph.ntu_rgb_d.Graph', 'dct': False, 'n_sample': 1, 'SAGC_proj': True, 'backbone': 'transformer', 'num_cls': 120, 'base_lr': 0.1, 'step': [50, 60], 'optimizer': 'SGD', 'nesterov': True, 'batch_size': 64, 'test_batch_size': 64, 'start_epoch': 0, 'num_epoch': 110, 'weight_decay': 0.0003, 'lr_decay_rate': 0.1, 'warm_up_epoch': 0, 'lambda_1': 1.0, 'lambda_2': 0.1, 'lambda_3': 0.1, 'lambda_4': 0.0, 'lambda_5': 0.0, 'half': True, 'amp_opt_level': 1, 'n_step': 3, 'n_min': 3, 'dilation': 1, 'ode_method': 'euler', 'work_dir': '/home/infoGCN--/wandb/run-20240102_065914-eeptaw4a/files'}
```

**Comment by adventureRyan** [comment]
I think the issue may be caused by your Python version being too high. I solved it by using Python 3.8.
Here are the details:
```
pyenv virtualenv 3.8 infogcn2
pyenv activate infogcn2
pip install torch==1.9.0+cu111 torchvision==0.10.0+cu111 torchaudio==0.9.0 -f https://download.pytorch.org/whl/torch_stable.html
pip install tensorboardX tqdm


pip install -e torchlight

pip install torchpack==0.2.2

pip install torch_dct einops torchdiffeq wandb matplotlib

cd apex
pip install -r requirements.txt
python setup.py install
```


And you might encounter an issue related to an Apex error.You can solve it by modify its source code.
For me, I changed `/home/ryan/.pyenv/versions/infogcn2/lib/python3.8/site-packages/apex-0.1-py3.8.egg/apex/transformer/tensor_parallel/mappings.py`
Change line 28 to the following format:
```
if "reduce_scatter_tensor" not in dir(torch.distributed):
    # torch.distributed.reduce_scatter_tensor = torch.distributed._reduce_scatter_base
    torch.distributed.reduce_scatter_tensor = torch.distributed.reduce_scatter
```

**Comment by Ant0ny44** [comment]
Thank you for your solution. It works now.
